staging/src/k8s.io/client-go/tools/leaderelection/resourcelock/multilock.go

```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resourcelock
+
+import (
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+)
+
+// MultiLock is used for lock's migration
+type MultiLock struct {
+	Primary   Interface
+	Secondary Interface
+}
+
+// Get returns the older election record of the lock
+func (ml *MultiLock) Get() (*LeaderElectionRecord, error) {
+	primary, err := ml.Primary.Get()
+	if err != nil {
+		return nil, err
+	}
+	secondary, err := ml.Secondary.Get()
+	if err != nil {
+		return nil, err
+	}
+	// check if the older resource lock is expired
```

Comment:
We have to be careful to avoid situations where a corrupt state (e.g. different leaders in each record causes unrecoverable errors). Another option is to change Get() signature to return `(record LeaderElectionRecord, rawRecord []byte, error)`, store the raw record as `le.observedRawRecord` and change this DeepEqual to a byte comparison:

https://github.com/kubernetes/kubernetes/blob/a7c81c663d4a59d0e4d9be96c394b7dad1a6cf76/staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go#L342-L346

We need a HolderIdentity that isn't "" to pass back from Get() when the two objects disagree on the leader disagree. We can pick a placeholder like "leaderelection.k8s.io/unknown". What we get is:
* transitional clients respect other clients that are using either the old or new resource only.
* transitional clients treat corruption as just another leader named "unknown" and will attempt to steal the lease after LeaseDuration.
* no dependencies on time.

This seems like a pretty minimal change to the leaderelection client, EndpointsLock and ConfigMapLock but LeaseLock is a bit tricky since we don't currently handle raw bytes for that record since it's stored in the object and not an annotation.

Reply:
No sure I'm really following what does rawRecord give us - DeepCopy should be doing the same stuff, right?

Reply:
The client detects a refresh of the the lease by checking if anything changed in the record. This is done with a DeepEqual. If we want a lock to only be valid if both primary and secondary are the same, transitional clients (moving to a new record) need to compare two locks while steady state clients need to compare only one lock. This is certainly possible to do:
* with DeepEqual by returning both records and comparing both in the client, then handling either one or two records everywhere after.
* with comparing both in the resource lock implementation, then returning an error. This would also require modifying the leaderelection client today to handle this kind of error and we need to be careful to make sure that corrupt states are recoverable.

I'm suggesting a third option:
* change the DeepEqual to be a bytes.Compare. Steady state clients return a single serialized record, transitional clients return two concatenated records.
* construct a LeaderElectionRecord struct in the transitional client picking the best fields to preserve all the informational logging.

The one field in LeaderElectionRecord that we use non-informationally is the Holder field. We use this to implement reentrance when a leader container restarts and to speed up the first election or elections after the leader explicty releases the lease. I client doesn't have too wait to reclaim the lease if the holder is empty or the holder == itself:

https://github.com/kubernetes/kubernetes/blob/d951d19ec4d8ec829e83d146200ea91b664b6fa6/staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go#L347-L349

I'm not sure this behavior is actually useful but it's there for now. The client should work correctly but a little slower without these optimizations. This is why I suggested that transitional clients treat mismatched HolderIdentities between the two records as just another leader named "unknown" and will attempt to steal the lease after LeaseDuration. This would be functionally equivalent to a steady state client reading a LeaderElectionRecord owned by another client (assuming the client isn't named "unknown" :)). We are just skipping the HolderIdentity optimizations.

Reply:
> I'm suggesting a third option:
change the DeepEqual to be a bytes.Compare. Steady state clients return a single serialized record, transitional clients return two concatenated records.

OK - that makes a lot of sense to me.

Re holder-identity - yes, that I also already agreed with before.

Reply:
> The one field in LeaderElectionRecord that we use non-informationally is the Holder field. We use this to implement reentrance when a leader container restarts and to speed up the first election or elections after the leader explicty releases the lease. I client doesn't have too wait to reclaim the lease if the holder is empty or the holder == itself:
> 

I think the optimization is not work while restarting kcm and scheduler. Holder is generated with hostname and uuid, so the old holder will not equal to the new one.

Reply:
Yup, it doesn't get hit e.g. during a master upgrade or a master restart.